File: common/src/main/java/me/kvalbrus/multibans/common/punishment/MultiPermanentlyPunishment.java
package me.kvalbrus.multibans.common.punishment;

import java.util.List;
import me.kvalbrus.multibans.api.punishment.PermanentlyPunishment;
import me.kvalbrus.multibans.api.punishment.creator.PunishmentCreator;
import me.kvalbrus.multibans.api.punishment.PunishmentStatus;
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget;
import me.kvalbrus.multibans.api.punishment.PunishmentType;
import me.kvalbrus.multibans.common.managers.PluginManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class MultiPermanentlyPunishment extends MultiPunishment
    implements PermanentlyPunishment {

    private PunishmentCreator cancellationCreator;

    private long cancellationDate;

    private String cancellationReason;

    private boolean cancelled;

    public MultiPermanentlyPunishment(@NotNull PluginManager pluginManager,
                                      @NotNull PunishmentType type,
                                      @NotNull String id,
                                      @NotNull PunishmentTarget target,
                                      @NotNull PunishmentCreator creator,
                                      long dateCreated,
                                      @Nullable String reason,
                                      @Nullable String comment,
                                      @Nullable PunishmentCreator cancellationCreator,
                                      long cancellationDate,
                                      @Nullable String cancellationReason,
                                      @NotNull List<String> servers,
                                      boolean cancelled) {
        super(pluginManager, type, id, target, creator, dateCreated, reason, comment, servers);
        this.cancellationCreator = cancellationCreator;
        this.cancellationDate = cancellationDate;
        this.cancellationReason = cancellationReason;
        this.cancelled = cancelled;
    }

    @Override
    public final synchronized void activate() {
        this.cancelled = false;
        super.activate();
    }

    @Override
    public final synchronized void deactivate(@Nullable PunishmentCreator cancellationCreator,
                                        long cancellationDate,
                                        @Nullable String cancellationReason) {
        if (this.cancelled) {
            return;
        }

        this.cancelled = true;
        this.cancellationCreator = cancellationCreator;
        this.cancellationDate = cancellationDate;
        this.cancellationReason = cancellationReason;

        this.updateData();
        this.sendMessageAboutDeactivate();
    }

    @Override
    public final synchronized void delete() {
        super.delete();
    }

    @NotNull
    @Override
    public PunishmentStatus getStatus() {
        if (this.cancelled) {
            return PunishmentStatus.CANCELLED;
        } else {
            return PunishmentStatus.ACTIVE;
        }
    }

    @Nullable
    @Override
    public PunishmentCreator getCancellationCreator() {
        return this.cancellationCreator;
    }

    @Override
    public long getCancellationDate() {
        return this.cancellationDate;
    }

    @Nullable
    @Override
    public String getCancellationReason() {
        return this.cancellationReason;
    }

    @Override
    public boolean isCancelled() {
        return this.cancelled;
    }

    @Override
    public synchronized void setCancellationCreator(@Nullable PunishmentCreator cancellationCreator) {
        this.cancellationCreator = cancellationCreator;
        this.updateData();
    }

    @Override
    public synchronized void setCancellationDate(long cancellationDate) {
        this.cancellationDate = cancellationDate;
        this.updateData();
    }

    @Override
    public synchronized void setCancellationReason(@NotNull String cancellationReason) {
        this.cancellationReason = cancellationReason;
        this.updateData();
    }

    @Override
    public synchronized void setCancelled(boolean cancelled) {
        this.cancelled = cancelled;
        this.updateData();
    }

    @NotNull
    public abstract String getDeactivateMessageForListener();

    @NotNull
    public abstract String getDeactivateMessageForExecutor();

    @Nullable
    public abstract String getDeactivateMessageForTarget();

    private void sendMessageAboutDeactivate() {
        this.sendMessageToListeners(this.getDeactivateMessageForListener());
        this.sendMessageToCreator(this.getDeactivateMessageForExecutor());
        this.sendMessageToTarget(this.getDeactivateMessageForTarget());

    }
}

File: common/src/main/kotlin/me/kvalbrus/multibans/common/punishment/MultiPermanentlyPunishment.kt
package me.kvalbrus.multibans.common.punishment

import me.kvalbrus.multibans.api.punishment.PermanentlyPunishment
import me.kvalbrus.multibans.api.punishment.action.ActivationAction
import me.kvalbrus.multibans.api.punishment.action.CreationAction
import me.kvalbrus.multibans.api.punishment.action.DeactivationAction
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentStatus
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentType
import me.kvalbrus.multibans.api.punishment.executor.PunishmentExecutor
import me.kvalbrus.multibans.common.managers.PluginManager
import me.kvalbrus.multibans.common.punishment.action.MultiActivationAction
import me.kvalbrus.multibans.common.punishment.action.MultiDeactivationAction
import me.kvalbrus.multibans.common.punishment.punishments.ActionLoader

abstract class MultiPermanentlyPunishment : MultiPunishment, PermanentlyPunishment, ActionLoader {

    private var _activations: MutableList<ActivationAction> = mutableListOf()
    private var _deactivations: MutableList<DeactivationAction> = mutableListOf()
    private var _cancelled: Boolean

    constructor(
        pluginManager: PluginManager,
        type: PunishmentType,
        id: String,
        creationAction: CreationAction,
        comment: String,
        servers: List<String>,
        cancelled: Boolean) : super(pluginManager, type, id, creationAction, comment, servers) {
            this._cancelled = cancelled
    }

    constructor(
        pluginManager: PluginManager,
        type: PunishmentType,
        id: String,
        creationAction: CreationAction,
        activations: MutableList<ActivationAction>,
        deactivations: MutableList<DeactivationAction>,
        comment: String,
        servers: List<String>,
        cancelled: Boolean) : this(pluginManager, type, id, creationAction, comment, servers, cancelled) {
            this._activations = activations
            this._deactivations = deactivations
    }

    final override val activations: List<ActivationAction>
        get() = ArrayList(this._activations)

    final override val deactivations: List<DeactivationAction>
        get() = ArrayList(this._deactivations)

    final override val cancelled: Boolean
        get() = this._cancelled

    @Synchronized
    override fun activate(executor: PunishmentExecutor, date: Long, reason: String) : Boolean {
        return this.activate(MultiActivationAction(this.id, this._activations.size + 1, executor, date, reason))
    }

    @Synchronized
    override fun activate(action: ActivationAction) : Boolean {
        if (this._cancelled) {
            this._cancelled = false

            this._activations.add(action)
            this.updateData()
            this.pluginManager.activatePunishment(this)
            this.sendMessageAboutActivate()

            return true
        } else {
            return false
        }
    }

    @Synchronized
    override fun deactivate(executor: PunishmentExecutor, date: Long, reason: String): Boolean {
        return this.deactivate(MultiDeactivationAction(this.id, this._deactivations.size + 1, executor, date, reason))
    }

    @Synchronized
    override fun deactivate(action: DeactivationAction) : Boolean {
        if (!this._cancelled) {
            this._cancelled = true

            this._deactivations.add(action)
            this.updateData()
            this.pluginManager.deactivatePunishment(this)
            this.sendMessageAboutDeactivate()

            return true
        } else {
            return false
        }
    }

    override fun getStatus(): PunishmentStatus {
        return if (this._cancelled) {
            PunishmentStatus.CANCELLED
        } else {
            PunishmentStatus.ACTIVE
        }
    }

    override fun lastActivator(): PunishmentExecutor {
        if (this.activations.isEmpty()) {
            return this.creationAction.executor
        } else {
            var maxTimeIndex = 0
            val activations = this.activations
            for (i in 0..activations.size) {
                if (activations[maxTimeIndex].date < activations[i].date) {
                    maxTimeIndex = i
                }
            }

            return activations[maxTimeIndex].executor
        }
    }

    override fun lastDeactivator(): PunishmentExecutor? {
        if (this.deactivations.isEmpty()) {
            return null
        } else {
            var maxTimeIndex = 0
            val deactivations = this.deactivations
            for (i in 0..deactivations.size) {
                if (deactivations[maxTimeIndex].date < deactivations[i].date) {
                    maxTimeIndex = i
                }
            }

            return deactivations[maxTimeIndex].executor
        }
    }

    override fun loadActivationsFromDataProvider(activations: MutableList<ActivationAction>) {
        this._activations = activations
    }

    override fun loadDeactivationsFromDataProvider(deactivations: MutableList<DeactivationAction>) {
        this._deactivations = deactivations
    }

    private fun sendMessageAboutActivate() {
        sendMessageToListeners(this.activateMessageForListener)
        sendMessageToCreator(this.activateMessageForExecutor)
        sendMessageToTarget(this.activateMessageForTarget)
    }

    private fun sendMessageAboutDeactivate() {
        sendMessageToListeners(this.deactivateMessageForListener)
        sendMessageToCreator(this.deactivateMessageForExecutor)
        sendMessageToTarget(this.deactivateMessageForTarget)
    }

    abstract val activateMessageForListener: String
    abstract val activateMessageForExecutor: String
    abstract val activateMessageForTarget: String

    abstract val deactivateMessageForListener: String
    abstract val deactivateMessageForExecutor: String
    abstract val deactivateMessageForTarget: String
}